Add ability to add and remove peaks from SliceViewer

jmborr opened the pull request:
**Description of work.**

[SCD142](https://code.ornl.gov/sns-hfir-scse/diffraction/single-crystal/single-crystal-diffraction/-/issues/142)

**To test:**

Create MD workspaces (HKL and Q_sample) and peaks workspace, open with sliceviewer and overlay peaks, try adding and removing peaks.

```python
md_qsample = CreateMDWorkspace(Dimensions=3, Extents='0,10,0,10,0,10', Names='x,y,z', Units='r.l.u.,r.l.u.,r.l.u.', Frames='QSample,QSample,QSample')
md_hkl = CreateMDWorkspace(Dimensions=3, Extents='0,10,0,10,0,10', Names='H,K,L', Units='r.l.u.,r.l.u.,r.l.u.', Frames='HKL,HKL,HKL')

peaks = CreatePeaksWorkspace(OutputType='LeanElasticPeak', NUmberOfPeaks=0)
SetUB(peaks, a=2*np.pi, b=2*np.pi, c=2*np.pi, u='0,0,1', v='1,0,0')
```

You can also test with the non-orthogonal option on

```python
ws=CreateSampleWorkspace()
SetUB(ws, a=2*np.pi, b=2*np.pi, c=2*np.pi, u='0,0,1', v='1,0,0', Gamma=120)
md_hkl = CreateMDWorkspace(Dimensions=3, Extents='0,10,0,10,0,10', Names='H,K,L', Units='r.l.u.,r.l.u.,r.l.u.', Frames='HKL,HKL,HKL')
md_hkl.addExperimentInfo(ws)
peaks=CreatePeaksWorkspace(NumberOfPeaks=0, OutputType='LeanElasticPeak')
CopySample(ws, peaks, CopyEnvironment=False)
```

There is no associated issue.
---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

jmborr committed:
type hints and actions package

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

jmborr committed:
forward declarations to prevent import loops

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

jmborr committed:
forward declarations to prevent import loops

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

jenest committed:
Add buttons and combo box

jenest committed:
Update function name

jmborr committed:
subscribing the presenter to the viewer

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

jenest committed:
Modify grid layout

jenest committed:
Include correct import

jmborr committed:
Tying the PeaksActionModel to the presenter

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

jenest committed:
Add actions.ui file

jmborr committed:
Instantiating the peak actions

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

jmborr committed:
group box in ui file

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

jmborr committed:
append/remove PeaksWorkspace from the view

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

jmborr committed:
resolve circular imports

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

jmborr committed:
lazy connections between vier and presenter

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

jmborr committed:
getattr should've been hasattr

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

jmborr committed:
getattr should've been hasattr

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

jmborr committed:
route signals only when presenter is not None

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

rosswhitfield committed:
Hack to get everything showing, need to fix PeakAction size policy

rosswhitfield committed:
Get remove_peaks working

rosswhitfield committed:
Add a peak to the workspace

rosswhitfield committed:
Ugly but have peak adding working

rosswhitfield committed:
Add minimum size to PeakActionsView

rosswhitfield committed:
Move coordinate transform to sliceview presenter

rosswhitfield committed:
Move add_peak from PeakAction model to PeaksViewer model

rosswhitfield committed:
Trigger ADS update when adding peaks from python

jenest committed:
Check one button at a time

rosswhitfield committed:
Disable pan and zoom while adding peaks

jenest committed:
Add minimum width for actions widget

rosswhitfield committed:
Force draw after workspace replaced

rosswhitfield committed:
Deactivate peak adding if pan or zoom selected

jmborr committed:
Remove handle PeakActionsPresenter from PeaksViewerCollectionView

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

jmborr committed:
rename peaks_actions_view property

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

jmborr committed:
Revert "rename peaks_actions_view property"

This reverts commit 0b59766b913558323b8754f2970d815794e4f351.

jmborr committed:
Revert "Remove handle PeakActionsPresenter from PeaksViewerCollectionView"

This reverts commit c3eeb15b77a1cf4dfbfa4a29b385144b085a592e.

jmborr committed:
PeakActionsPresenter not used in PeaksViewerCollectionView

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

jmborr committed:
Remove actions presenter and view

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

jmborr committed:
move actions/view.py to peaksviewers/

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

rosswhitfield committed:
Cleanup peak adding

rosswhitfield committed:
Change 'button_press_event' connection

rosswhitfield committed:
Fix button selections

rosswhitfield committed:
Add tests for adding peaks in peaksviewer

rosswhitfield committed:
Add test for sliceviewer.presenter/peak_add_delete

jmborr committed:
peak deletion deletes peak closest to input position

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

jmborr committed:
Unify deleting/adding peaks

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

jmborr committed:
test delete peaks

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

rosswhitfield committed:
Don't need to include frame from sliceviewer, PeaksViewer knows this

rosswhitfield committed:
Change SliceViewer.peak_add_delete to add_delete_peak to make it the same as PeaksViewerCollectionPresenter

Also simplify PeaksViewerModel.delete_peak

rosswhitfield committed:
Add peaksviewer/sliceviewer integration test for add/remove peaks

jmborr commented on qt/python/mantidqt/widgets/sliceviewer/peaksviewer/model.py:
```diff
@@ -92,15 +92,12 @@ def add_peak(self, pos, frame):
+        if self.peaks_workspace.getNumberPeaks() == 0:
+            return
+
+        frame_to_slice_fn = self._frame_to_slice_fn(frame)
```
@rosswhitfield  I had noticed `_frame_to_slice_fn` but decided not to use it because I thought that self.peaks_workspace.column would be  faster and might be of some advantage when having thousands of peaks. Do you know if both approaches are equally fast?

rosswhitfield replied:
For 100000 peaks

```python
peaks.column('QSample')
```

takes ~0.04 seconds


```python
[peak.getQSampleFrame() for peak in peaks]
```

takes ~0.05 seconds

So yes it's about 25% slower. Do we care?

jmborr replied:
Just in case, I tried with `[getattr(peak, 'getQSampleFrame')() for peak in peaks]` but it only adds about another 0.01s so we'll leave it as it is

rosswhitfield replied:
By far the slowest part of all this is converting to a numpy array, I think it's because we are converting a list of V3D into a number array.

So for 100000 peaks

```python
np.array(peaks.column('QSample'))
```

takes about the same time (about 3 seconds) as

```python
np.array([getattr(peak, 'getQSampleFrame')() for peak in peaks])
```

If we do this instead

```python
np.array([list(getattr(peak, 'getQSampleFrame')()) for peak in peaks])
```

it cuts the time to about 1 second.

jmborr replied:
In that case we could use a trick we [introduced in the SANS backend](https://code.ornl.gov/sns-hfir-scse/sans/sans-backend/-/blob/next/drtsans/settings.py#L129). Running this in  my workstation:
```python
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np
import time

LoadEmptyInstrument(InstrumentName='TOPAZ', OutputWorkspace='topaz')
CreatePeaksWorkspace(InstrumentWorkspace='topaz', NumberOfPeaks=100000, OutputType='LeanElasticPeak', OutputWorkspace='peaks')
peaks = mtd['peaks']

start = time.time()
np.array(peaks.column('QSample'))
print('Elapsed column = ', time.time() - start)

start = time.time()
np.array([getattr(peak, 'getQSampleFrame')() for peak in peaks])
print('Elapsed getattr = ', time.time() - start)

start = time.time()
np.array([list(getattr(peak, 'getQSampleFrame')()) for peak in peaks])
print('Elapsed list getattr = ', time.time() - start)

def unpack_pos(peak):
    v3d_vector = getattr(peak, 'getQSampleFrame')()
    return [v3d_vector.X(), v3d_vector.Y(), v3d_vector.Z()]
    
start = time.time()
np.array([unpack_pos(peak) for peak in peaks])
print('Elapsed unpack_pos = ', time.time() - start) 
```

Results in:

> LoadEmptyInstrument started
> LoadEmptyInstrument successful, Duration 6.34 seconds
> CreatePeaksWorkspace started
> CreatePeaksWorkspace successful, Duration 1.32 seconds
> Elapsed column =  9.059121370315552
> Elapsed getattr =  9.599326133728027
> Elapsed list getattr =  4.169772148132324
> Elapsed unpack_pos =  1.5399858951568604

rosswhitfield replied:
Yeah probably can't get much faster than that unless we move this to c++. This could be implemented as a method on the PeaksWorkspace, but I don't know if we need to do that at this point.

jmborr committed:
send AfterReplaceNotification after removal of a peak

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

jmborr commented:
@rosswhitfield Since Pete's shipped you the visibility task, this is what I think needs to be done:  

I think we only need to amend `PeaksViewerCollectionView` methods `append_peaksviewer` and `remove_peaksviewer`. 

```python
def append_peaksviewer(..):
    . . .
    self._peak_actions_view.setVisible(True)
```

```python
def remove_peaksviewer(..):
    . . .
    (pseudocode) if no more PeaksViewerView objects in self._peaks_layout:
        self._peak_actions_view.setVisible(False)

rosswhitfield committed:
Fix coordinate transform

rosswhitfield committed:
Make PeaksViewerModel.delete_peak faster

rosswhitfield committed:
Fix test_peak_add_delete_event

rosswhitfield committed:
Fix workspace ordering with multiple peaks workspaces

And cleanup some unused methods.

rosswhitfield committed:
Add test for PeakActionView.set_peaksworkspace

rosswhitfield committed:
Update docs

rosswhitfield committed:
Change replace/rename handle to keep same position

rosswhitfield committed:
Adjust test for changes to replace/rename handle

jenest committed:
Set visibility specs

rosswhitfield committed:
Update docs

rosswhitfield committed:
Forgot to add new image

zjmorgan commented:
Not sure if this is is related to the added functionality, but I tried adding peaks then switching axes. That caused the following exception causing Mantid to close
![image](https://user-images.githubusercontent.com/13754794/120022324-42c59780-bfba-11eb-99d9-72efe9a83ce2.png)

zjmorgan commented:
The above issue only seems to be a problem when adding peaks to HKL workspace.

rosswhitfield commented:
> Not sure if this is is related to the added functionality, but I tried adding peaks then switching axes. That caused the following exception causing Mantid to close
> ![image](https://user-images.githubusercontent.com/13754794/120022324-42c59780-bfba-11eb-99d9-72efe9a83ce2.png)

This bug is unrelated to this PR. You can reproduce it with the mantidworkbenchnightly with

```python
md_hkl = CreateMDWorkspace(Dimensions=3, Extents='0,10,0,10,0,10', Names='H,K,L', Units='r.l.u.,.r.l.u.,r.l.u.', Frames='HKL,HKL,HKL')
peaks = CreatePeaksWorkspace(OutputType='LeanElasticPeak', NUmberOfPeaks=1)
```

It also only happens when viewing H and L and not any other combination.

rosswhitfield commented:
So the cause of that issue is an extra `.` in the units that makes it think that it's not a q dimension.

```diff
< md_hkl = CreateMDWorkspace(Dimensions=3, Extents='0,10,0,10,0,10', Names='H,K,L', Units='r.l.u.,.r.l.u.,r.l.u.', Frames='HKL,HKL,HKL')
---
> md_hkl = CreateMDWorkspace(Dimensions=3, Extents='0,10,0,10,0,10', Names='H,K,L', Units='r.l.u.,r.l.u.,r.l.u.', Frames='HKL,HKL,HKL')
```

I fixed the test instructions.

zjmorgan commented:
Correcting that made this issue go way selecting peaks in the HL plane by selecting it first before opening the peaks table 
```
Traceback (most recent call last):
  File "/usr/lib64/python3.6/site-packages/matplotlib/cbook/__init__.py", line 387, in process
    proxy(*args, **kwargs)
  File "/usr/lib64/python3.6/site-packages/matplotlib/cbook/__init__.py", line 227, in __call__
    return mtd(*args, **kwargs)
  File "/SNS/users/zgf/.git/mantid/qt/python/mantidqt/widgets/sliceviewer/presenter.py", line 426, in add_delete_peak
    pos = sliceinfo.inverse_transform([event.xdata, event.ydata, sliceinfo.z_value])
  File "/SNS/users/zgf/.git/mantid/qt/python/mantidqt/widgets/sliceviewer/sliceinfo.py", line 103, in inverse_transform
    point = np.dot(inv_trans, point)
TypeError: unsupported operand type(s) for *: 'float' and 'NoneType'
```

zjmorgan commented:
The added functionality all seems to work as expected. Adding and deleting peaks works (intuitively) well. Verified that the both LeanElasticPeaks and Peaks workspace work both in MD HKL and Q space.

zjmorgan commented on docs/source/workbench/sliceviewer.rst:
```diff
@@ -44,11 +63,35 @@ Peaks are displayed at the locations defined by each peak center
+   :width: 75%
+   :align: center
+
+Adding or removing peaks
```
Documentation looks good.

zjmorgan commented on qt/python/mantidqt/widgets/sliceviewer/peaksviewer/test/test_peaksviewer_sliceviewer_add_delete_peaks_integration.py:
```diff
@@ -0,0 +1,205 @@
+        md = CreateMDWorkspace(Dimensions=3,
+                               Extents='0,10,0,10,0,10',
+                               Names='H,K,L',
+                               Units='r.l.u.,.r.l.u.,r.l.u.',
```
Doesn't look like `.r.l.u.` will hurt the HKL test

zjmorgan commented:
Happy to approve this but I think there should be one additional reviewer.

rosswhitfield committed:
Fix two issue with peak overlay display,

Should be checking all workspaces instead of just newly added workspaces. Deactivate peak adding/removing when peak overlay is closed

zjmorgan reviewed (approved):
All looks good.
